### PR TITLE
Fixed missing invocation of Address Sanitizer

### PR DESCRIPTION
**Pull request by davidbolvansky**
Unfortunely invocation of ASan was not working as intended because check_cxx_compiler_flag (called by add_cxx_compiler_option) doesn't give correct result for -fsanitize=address.

https://cmake.org/pipermail/cmake/2019-October/070179.html

Fix: add flags directly, do not use add_cxx_compiler_option in this case.

GCC then crashed when building with ASAN with "c++: fatal error: Killed signal terminated program cc1plus" so use sanitizers with Clang build.

### CODE REVIEW

File: backends/ubpf/ubpfTable.cpp [hunk 1]
```diff
@@ -191,24 +191,24 @@ void UBPFTable::emitInstance(EBPF::CodeBuilder* builder) {
 }
 
 void UBPFTable::setTableKind() {
-    if (keyGenerator == nullptr) {
-        return;
```
Comment (davidbolvansky) [inline backends/ubpf/ubpfTable.cpp:195]: In this case this->tableKind was not initialized.

File: lib/cstring.cpp [hunk 2]
```diff
@@ -187,10 +187,12 @@ cstring cstring::newline = cstring("\n");
 cstring cstring::empty = cstring("");
 
 bool cstring::startsWith(const cstring& prefix) const {
+    if (prefix.isNullOrEmpty()) return true;
```
Comment (davidbolvansky) [inline lib/cstring.cpp:190]: memcmp should not be called with NULL pointers.

File: tools/ci-build.sh [hunk 3]
```diff
@@ -177,6 +177,12 @@ CMAKE_FLAGS+="-DENABLE_SANITIZERS=${ENABLE_SANITIZERS} "
 # Enable auto var initialization with pattern.
 CMAKE_FLAGS+="-DBUILD_AUTO_VAR_INIT_PATTERN=${BUILD_AUTO_VAR_INIT_PATTERN} "
 
+if [ "$ENABLE_SANITIZERS" == "ON" ]; then
+  export UBSAN_OPTIONS=print_stacktrace=1
+  export ASAN_OPTIONS=print_stacktrace=1:detect_leaks=0
+  CMAKE_FLAGS+="-DENABLE_GC=OFF"
```
Comment (fruffy) [inline tools/ci-build.sh:183]: Maybe add a warning that gc is turned off here.

**Review by fruffy — APPROVED** [review]

File: .github/workflows/ci-test.yml [hunk 4]
```diff
@@ -63,16 +63,17 @@ jobs:
...
+        docker build -t p4c --build-arg IMAGE_TYPE=test --build-arg ENABLE_UNIFIED_COMPILATION=ON --build-arg COMPILE_WITH_CLANG=ON \
+        --build-arg BUILD_AUTO_VAR_INIT_PATTERN=ON --build-arg ENABLE_SANITIZERS=ON .
         ./tools/export_ccache.sh
 
       # run with sudo (...) --privileged
       # this is needed to create network namespaces for the ebpf tests.
     - name: Run tests (Ubuntu Linux)
       run: |
-        sudo docker run --privileged -w /p4c/build -e $CTEST_PARALLEL_LEVEL p4c ctest --output-on-failure --schedule-random
+        sudo docker run --privileged -w /p4c/build -e $CTEST_PARALLEL_LEVEL -e UBSAN_OPTIONS=print_stacktrace=1 -e ASAN_OPTIONS=print_stacktrace=1:detect_leaks=0 p4c ctest --output-on-failure --schedule-random
```
[the remark this answers is not in the document]
Comment (davidbolvansky) [reply .github/workflows/ci-test.yml:76]: IIRC due to irgenerator

File: backends/bmv2/CMakeLists.txt [hunk 5]
```diff
@@ -190,11 +190,14 @@ set (XFAIL_TESTS
 set (BMV2_PARSER_INLINE_TESTS "${P4C_SOURCE_DIR}/testdata/p4_16_samples/parser-inline/*.p4")
 
 if (HAVE_SIMPLE_SWITCH)
-  p4c_add_tests("bmv2" ${BMV2_DRIVER} "${BMV2_V1MODEL_TEST_SUITES}" "${XFAIL_TESTS}")
-  add_library(extern_func_module MODULE EXCLUDE_FROM_ALL "${P4C_SOURCE_DIR}/testdata/extern_modules/extern-funcs-bmv2.cpp" )
-  p4c_add_test_with_args("bmv2" ${BMV2_DRIVER} FALSE "bmv2_emit_externs" "testdata/p4_16_samples/extern-funcs-bmv2.p4" "-a=--emit-externs --target-specific-switch-arg=\"--load-modules ${CMAKE_CURRENT_BINARY_DIR}/libextern_func_module.so\" --init=\"make extern_func_module\"" "")
-  p4c_add_tests("bmv2-parser-inline-opt-disabled" ${BMV2_DRIVER} "${BMV2_PARSER_INLINE_TESTS}" "")
-  p4c_add_tests("bmv2-parser-inline-opt-enabled" ${BMV2_DRIVER} "${BMV2_PARSER_INLINE_TESTS}" "" "-a=--parser-inline-opt")
+  if (NOT ENABLE_SANITIZERS)
```
[the remark this answers is not in the document]
Comment (davidbolvansky) [reply backends/bmv2/CMakeLists.txt:193]: HAVE_SIMPLE_SWITCH AND NOT ENABLE_SANITIZERS? 

but then warning message in else branch would be incorrect

File: backends/ubpf/ubpfTable.cpp (hunk 1 above)
Comment (davidbolvansky) [reply backends/ubpf/ubpfTable.cpp:195]: Will do.

File: backends/bmv2/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (fruffy) [reply backends/bmv2/CMakeLists.txt:193]: True, let's keep it this way then.

**Review by fruffy — APPROVED** [review]

**Pull request merged**